### ISSUE TITLE
Alerts not logged

### ISSUE THREAD

**Issue by MumpekPumpa** [BinaryDefense/artillery#86]
I'm running 2.1.1 (also tried the newest version) on Windows Server 2019. I start artillery, the log file alerts.log confirms this with "Artillery has started successfully", then I do a nmap scan, I get opened ports that were configured, but nothing gets logged. MONITOR &  CONSOLE LOGGING are both ON. The server has two network cards, but I don't get any alerts even if one is disabled.

Any ideas?

**Comment by russhaun** [comment]
Hey just saw this post. I'm not at my lab right now. Do you get alerts on screen and not in logs.? Have you tried setting the interface manually  in configuration file?. I will be home later today and will spin up an instance of 2019 and test for you to see if I can replicate this issue. I just need py ver

**Comment by MumpekPumpa** [comment]
Thank you for replying, no I don't get anything on the screen either. I don't know what changed, since it was working fine weeks ago, now nothing. I uninstalled, reinstalled several times.
No I didn't try setting it manually (didn't know it was possible and how), py version is 3.7.3.

**Comment by russhaun** [comment]
In v 2.1.1 there is an option "bind_interface" located in the configuration file. Set this to an available ip. Because you said it worked a little bit ago try disabling firewall and do another scan see if it picks it up. Also check your routing table and make sure the ip you are scanning from is not present

**Comment by russhaun** [comment]
i am installing 2019 now i will update you when i have more. you said you have 2 nics in this system. are they seperate ips or on same network?

**Comment by MumpekPumpa** [comment]
@russhaun thank you for your assistance, I performed additional tests with nmap and I only get alerts (on screen and logs) when I perform "Intense scans". If I only test a particular port that I know is open I don't get any logging, same with "Regular scan".

Is that by design? nmap still show the port is open, but nothing gets logged?

**Comment by corelanc0d3r** [comment]
I believe python only sees a connection if a full handshake is done - if you're doing a stealth scan, python may not see it.   (which may be good or bad... in any case, if a connection is detected, you're pretty sure it's someone trying really "connect“ :) )

**Comment by russhaun** [comment]
@MumpekPumpa 

just a small update. :)

i can confirm artillery does run on 2019. but not without issues. 2.1.1 will run the issues occur when it gets scanned. from what i see nmap does a full connect when scanning which forcibly closes the connection producing errors on screen.  it does log conn. and adds to routing table. This is just doing ex: nmap <ip>. adding switches just adds to the errors on screen. Doing a stealth scan does not trigger artillery.i think we would want to catch this? this is a base install of 2019 and 2.1.1.

![issue_#86(srv 2019)](https://user-images.githubusercontent.com/20805369/71602237-5a7ac100-2b25-11ea-99be-3d8ac7aca317.png)


On a base 2019 and (current release) the number of errors is increased and also some issues with imports for email. it does not log or add to routing table. it never gets that far in the code. 

To the issue of dual nics. you can only bind to one nic in the current implementation.(socmiiw please) if you try both it will fail it will choose one on start and ignore the other. also there are issues with the udp server windows complains. it can be done from what i've researched but it will have to be re-written to support it.
             
Now with all that said i do have a fix in the works for the(current release). certain config checks are not being performed on startup causing it to fail. with the checks in place it works as intended. you do still see conn err though(working on it)  
![issue#86(current release_fixed)](https://user-images.githubusercontent.com/20805369/71602768-c2caa200-2b27-11ea-83b3-ba70e26c20fc.png)
if you made it this far and you would like to test it out let me know i can make a branch from my repo for you

**Comment by corelanc0d3r** [comment]
does the config file get created when you run artillery.py for the first time?
(I haven't checked setup.py - perhaps a call to the "check_config() (src/core.py)" in setup.py may fix that issue?)

**Comment by russhaun** [comment]
Hey @corelanc0d3r   yeah no file is created in either case setup/artillery.py.it does install but it complains about imports and then closes. your current pull request should fix it. just waiting on that to get merged. all of my changes will be in artillery.py and the new config section in core.py. figured its just better to wait for that to be applied. just need to set appropriate defaults for (windows) as some of the linux features are not implemented in windows yet so they dont need to default to "ON".

**Comment by corelanc0d3r** [comment]
pull request got merged.  Curious to see what options you will be changing - Just thinking out loud: maybe there is a need to indicate what options are not implemented on windows yet (rather than changing their default value. Changing default value would still not affect windows, but it might negatively affect other OSes)

**Comment by russhaun** [comment]
hey @corelanc0d3r just tried out the new release. still have issues. now it fails at trying to import globals here is a traceback.

C:\Users\russ\Downloads\artillery-master>python setup.py
Traceback (most recent call last):
  File "setup.py", line 11, in <module>
    from src.core import *
  File "C:\Users\russ\Downloads\artillery-master\src\core.py", line 48, in <module>
    import globals
ModuleNotFoundError: No module named 'globals'

i disagree on changing values. all values that were changed(current release) negativley affected windows and prevented it from running. also certain config checks are not being done so they run regardless of os which it shouldn't do. so if checks were in place for each os appropriate values will be selected for each os as it should be.

**Comment by corelanc0d3r** [comment]
wow - that's odd - does the globals.py file exist on your system (should be in the src folder)? 

regarding the changes - I was just curious, of course options that prevent artillery from running should either be fixed or disabled....   I can certainly help checking why they break artillery, perhaps we build fixes for them one by one

**Comment by russhaun** [comment]
little update:
             yes the globals.py file exists. jfsag i fired up my win7/py27 vm and it worked. here is a screen shot with items that need to be changed for windows. they dont need to start.
![issue#_86_py27](https://user-images.githubusercontent.com/20805369/71997399-d18e1480-320b-11ea-8b87-f3d4533a43a3.png)

this seems to be a py 3 issue with the import of the globals file on windows. thats weird

**Comment by russhaun** [comment]
hey @MumpekPumpa 
      update;
            still working on a fix for py 3 there is an issue with imports right now.artillery does run on py 2.7 so for now if you want to use it you can run it that way. you can disable any uneeded items through the config once it is installed. will let you know when i have more

**Comment by corelanc0d3r** [comment]
ok, I guess we can do a check for those services and have artillery ignore the setting if the detected OS is Windows. That way, it wouldn't matter if we enable or disable them. (I prefer to keep them enabled by default). Would that be ok?

regarding python3... not sure what the issue is :( I'll be traveling until the end of next week, I'll see if I can set up a windows box to do more testing...

**Comment by russhaun** [comment]
hey @corelanc0d3r 
           this is what i mean by checks not being performed. below you can see the ftp check but none of the 3 below have checks made. even if you set them they are ignored and also the harden var is not in config section so that would have to be added. you can default to on no big deal i can just turn it off.
![issue#86_config_checks](https://user-images.githubusercontent.com/20805369/72004397-4ddb2480-3219-11ea-9b0b-a92e18b61069.png)

**Comment by corelanc0d3r** [comment]
ok, I have added the checks in PR 91, and I believe i have also fixed the import issue on python3 - let's wait for the fix to be merged

**Comment by corelanc0d3r** [comment]
ok, made a few more changes, please check PR91

**Comment by corelanc0d3r** [comment]
@russhaun PR has been merged, can you check if things look better now?
Also, if the email mechanism breaks on windows, would it be possible to open a new issue for that case (and perhaps post some error information), so I can try to fix it ?
thanks !

**Comment by russhaun** [comment]
hey @corelanc0d3r :
 just downloaded the latest. it installs but fails to run due to permissions. i have to launch an elevated promt before i can run it. it never runs the uac script it launches but then gives me this.
![issue#86(email issue)](https://user-images.githubusercontent.com/20805369/72107432-5148db80-32ff-11ea-89a3-19902a78eb08.png)

so i forked it and made a new branch.please look at it here

https://github.com/russhaun/artillery/tree/win_fixes

which when run now acts as it should with perms and email.please look at commit message
along with suggested changes and produces this. just hade to move a few things around

![issue#86(fixes_applied)](https://user-images.githubusercontent.com/20805369/72107850-404c9a00-3300-11ea-9a2b-935521556eb7.png)


as long as your good with my changes i can ask for PR. this is all wip so if any changes need to be made we can get it all worked out here

tested on win 10 w py 3.7.3

**Comment by russhaun** [comment]
hey @corelanc0d3r 
update

ran my branch on kali latest with py 2.7 and 3.7.5 this is what i get. have to populate the smtp username and password value to make it go away on windows  even after that on linux this is what i get

![issue#86(linux_email)](https://user-images.githubusercontent.com/20805369/72115271-78f66e80-3314-11ea-82a1-9456b6e624fd.png)

**Comment by corelanc0d3r** [comment]
hey Russ,

I'm going through your changes and most things look ok

I do believe we should not prevent email from working if no username is provided.  An empty username doesn't mean the email handler shouldn't work. It only means the handler should not attempt to authenticate and do STARTTLS, but it should still try to send email.   (for instance, on my linux servers I am simply using a localhost relay that then authenticates somewhere else.)

Also, maybe I'm misreading the code, but it looks like you've moved check_config() entirely into the "if is_windows()"  branch, which means it won't be called anymore on posix. (it looks like the indentation of the "else“ is linked to "if isUserAdmin".  I have left a comment in your commit so you can check.
In any case, check_config() must always be executed (unless you're not running as admin/root)

thanks !

Peter

**Comment by corelanc0d3r** [comment]
the email issue on your machine seems to be an authentication issue - incorrect username/password
let's still print the error, but maybe it makes sense to see if we can make the code continue to run.  (maybe we need to wrap some code into another try/except block)

**Comment by russhaun** [comment]
Hey @corelanc0d3r 
       yeah i just setup the email issue to fire so you can see it. i agree on it being enabled.maybe we can put in some kind of check for it ?. i moved the check_config statement for both windows/posix to be fired after user is found to be admin only because you need the perms to look at the dirs associated with artillery. the entry for linux is @ line 67 of artillery.py.

yeah i figured as much on the auth issue just wanted to make you aware. im gonna submit this PR now.

**Comment by russhaun** [comment]
hey @MumpekPumpa 
 just want to let you know where im @ with this so far. right now from master(latest) this is what i get
![issue#86(server 2019_ routing add fail)](https://user-images.githubusercontent.com/20805369/72175533-5dd63e00-33aa-11ea-8f4e-c86cce571d7a.png)

from this you can see it starts and detects an attack but never adds it to routing table. while it starts and detects the function i highlighted causes it to fail before it gets to that point. specifically because of the format(err) function that is inside of write_log(). on windows it does not like that and goes duh and gives me a TypeError.

i do have a fix in the works. i just have to do a little conversion of things and then it will complete and add to routing table. also once this is resolved the other issues related to https certs will go away as well now because it can log and continue.

**Comment by russhaun** [comment]
hey @MumpekPumpa
        here is my latest result. all logging ang routing table stuff works now .my only other issue is now the banlist is not made properly. working on that
![isssue#86(writelog)](https://user-images.githubusercontent.com/20805369/72193667-7cebc480-33d8-11ea-94bd-740d29372464.png)

please check out this branch and tell me what you get when run.

https://github.com/russhaun/artillery/tree/writelog%2386

i can always clean up the screen just want to make sure it runs for you

**Comment by russhaun** [comment]
hey @MumpekPumpa 

 just asked for a PR to resolve this issue for good. with it all logging and alerting works now. routing add function works. now the nmap thing. this one is interesting in the results i got. with multiple interfaces i found that the same scan performed from different host depending on the interface it gave different results?. i'm gonna have to look into this. maybe i can put something together targeting nmap specifically to see what the issue is.

with that said there is one more issue to address. the banlist. while logging and banning works. the initial creation of banlist is broken right now it fails after adding the header at top of file? but it will append any new entries. will update you on that after i find the issue.

**Comment by russhaun** [comment]
PR was accepted all issues were resolved. you can close this now
